internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -170,6 +170,9 @@ func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
 }
 
 func (r *delegatingResolver) Close() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
```

Comment:
This is not sufficient to avoid the panic in the issue.  It will serialize the calls, but `Close` still lock the mutex before `updateTargetResolverState` and `updateTargetResolverState` will panic.

I have a possible fix [here](https://github.com/grpc/grpc-go/compare/master...arjan-bal:fix-delegating-resolver-race?expand=1). I'm working on adding a test that verifies this panic doesn't happen.